File: addons/guide-cpp/src/remapping/guide_remapping_config.cpp
#include "guide_remapping_config.h"

using namespace godot;

void GUIDERemappingConfig::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_remapped_inputs"), &GUIDERemappingConfig::get_remapped_inputs);
    ClassDB::bind_method(D_METHOD("set_remapped_inputs", "inputs"), &GUIDERemappingConfig::set_remapped_inputs);
    ADD_PROPERTY(PropertyInfo(Variant::DICTIONARY, "remapped_inputs"), "set_remapped_inputs", "get_remapped_inputs");

    ClassDB::bind_method(D_METHOD("get_custom_data"), &GUIDERemappingConfig::get_custom_data);
    ClassDB::bind_method(D_METHOD("set_custom_data", "data"), &GUIDERemappingConfig::set_custom_data);
    ADD_PROPERTY(PropertyInfo(Variant::DICTIONARY, "custom_data"), "set_custom_data", "get_custom_data");
}

GUIDERemappingConfig::GUIDERemappingConfig() {
}

GUIDERemappingConfig::~GUIDERemappingConfig() {
}

void GUIDERemappingConfig::_bind(const Ref<GUIDEMappingContext> &mapping_context, const Ref<GUIDEAction> &action, const Ref<GUIDEInput> &input, int index) {
    if (!remapped_inputs.has(mapping_context)) {
        remapped_inputs[mapping_context] = Dictionary();
    }
    Dictionary context_dict = remapped_inputs[mapping_context];
    if (!context_dict.has(action)) {
        context_dict[action] = Dictionary();
    }
    Dictionary action_dict = context_dict[action];
    action_dict[index] = input;
}

void GUIDERemappingConfig::_unbind(const Ref<GUIDEMappingContext> &mapping_context, const Ref<GUIDEAction> &action, int index) {
    _bind(mapping_context, action, Ref<GUIDEInput>(), index);
}

void GUIDERemappingConfig::_clear(const Ref<GUIDEMappingContext> &mapping_context, const Ref<GUIDEAction> &action, int index) {
    if (!remapped_inputs.has(mapping_context)) return;
    Dictionary context_dict = remapped_inputs[mapping_context];
    if (!context_dict.has(action)) return;
    Dictionary action_dict = context_dict[action];
    action_dict.erase(index);
    if (action_dict.is_empty()) {
        context_dict.erase(action);
    }
    if (context_dict.is_empty()) {
        remapped_inputs.erase(mapping_context);
    }
}

Ref<GUIDEInput> GUIDERemappingConfig::_get_bound_input_or_null(const Ref<GUIDEMappingContext> &mapping_context, const Ref<GUIDEAction> &action, int index) const {
    if (!remapped_inputs.has(mapping_context)) return Ref<GUIDEInput>();
    Dictionary context_dict = remapped_inputs[mapping_context];
    if (!context_dict.has(action)) return Ref<GUIDEInput>();
    Dictionary action_dict = context_dict[action];
    return action_dict.get(index, Ref<GUIDEInput>());
}

bool GUIDERemappingConfig::_has(const Ref<GUIDEMappingContext> &mapping_context, const Ref<GUIDEAction> &action, int index) const {
    if (!remapped_inputs.has(mapping_context)) return false;
    Dictionary context_dict = remapped_inputs[mapping_context];
    if (!context_dict.has(action)) return false;
    Dictionary action_dict = context_dict[action];
    return action_dict.has(index);
}

File: addons/GuideCS/Remapping/GUIDERemappingConfig.cs
using Godot;
using Godot.Collections;

namespace Guide;

[GlobalClass, Icon("res://addons/GuideCS/GUIDEInternal.svg")]
public partial class GUIDERemappingConfig : Resource {
	[Export]
	public Dictionary remapped_inputs = new();

	[Export]
	public Dictionary custom_data = new();

	internal void _Bind(GUIDEMappingContext mappingContext, GUIDEAction action, GUIDEInput input, int index = 0) {
		if (!remapped_inputs.ContainsKey(mappingContext)) {
			remapped_inputs[mappingContext] = new Dictionary();
		}

		var contextDict = (Dictionary)remapped_inputs[mappingContext];
		if (!contextDict.ContainsKey(action)) {
			contextDict[action] = new Dictionary();
		}

		var actionDict = (Dictionary)contextDict[action];
		actionDict[index] = input;
	}

	internal void _Unbind(GUIDEMappingContext mappingContext, GUIDEAction action, int index = 0) {
		_Bind(mappingContext, action, null, index);
	}

	internal void _Clear(GUIDEMappingContext mappingContext, GUIDEAction action, int index = 0) {
		if (!remapped_inputs.ContainsKey(mappingContext)) return;

		var contextDict = (Dictionary)remapped_inputs[mappingContext];
		if (!contextDict.ContainsKey(action)) return;

		var actionDict = (Dictionary)contextDict[action];
		actionDict.Remove(index);

		if (actionDict.Count == 0) {
			contextDict.Remove(action);
		}

		if (contextDict.Count == 0) {
			remapped_inputs.Remove(mappingContext);
		}
	}

	internal GUIDEInput _GetBoundInputOrNull(GUIDEMappingContext mappingContext, GUIDEAction action, int index = 0) {
		if (!remapped_inputs.ContainsKey(mappingContext)) return null;

		var contextDict = (Dictionary)remapped_inputs[mappingContext];
		if (!contextDict.ContainsKey(action)) return null;

		var actionDict = (Dictionary)contextDict[action];
		return actionDict.ContainsKey(index) ? actionDict[index].As<GUIDEInput>() : null;
	}

	internal bool _Has(GUIDEMappingContext mappingContext, GUIDEAction action, int index = 0) {
		if (!remapped_inputs.ContainsKey(mappingContext)) return false;

		var contextDict = (Dictionary)remapped_inputs[mappingContext];
		if (!contextDict.ContainsKey(action)) return false;

		var actionDict = (Dictionary)contextDict[action];
		return actionDict.ContainsKey(index);
	}
}
